app/ui/src/main/java/de/mm20/launcher2/ui/launcher/search/common/SearchableItemVM.kt

```diff
@@ -170,4 +182,114 @@ class SearchableItemVM : ListItemViewModel(), KoinComponent {
+    }
+
+    fun startHeadingUpdates(context: Context) {
+        if (trueNorthHeading.value != null) // somebody is listening!!! O_o
+            return
+
+        context
+            .getSystemService<SensorManager>()
+            ?.runCatching {
+                this.registerListener(
+                    sensorListener,
+                    this.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) ?: return@runCatching,
+                    SensorManager.SENSOR_DELAY_UI
+                )
+            }?.onFailure {
+                Log.e("SearchableItemVM", "Failed to start heading updates", it)
+            }
+    }
+
+    fun startLocationUpdates(context: Context) {
+        if (userLocation.value != null)
+            return
+
+        context
+            .getSystemService<LocationManager>()
+            ?.runCatching {
+                val hasFineAccess =
+                    context.checkPermission(Manifest.permission.ACCESS_FINE_LOCATION)
+                val hasCoarseAccess =
+                    context.checkPermission(Manifest.permission.ACCESS_COARSE_LOCATION)
+
+                userLocation.value =
+                    (if (hasFineAccess) this.getLastKnownLocation(LocationManager.GPS_PROVIDER) else null)
+                        ?: if (hasCoarseAccess) this.getLastKnownLocation(LocationManager.NETWORK_PROVIDER) else null
+
+                if (userLocation.value != null) {
+                    geomagneticField = GeomagneticField(
+                        userLocation.value!!.latitude.toFloat(),
+                        userLocation.value!!.longitude.toFloat(),
+                        userLocation.value!!.altitude.toFloat(),
+                        userLocation.value!!.time
+                    )
+                }
+
+                if (hasFineAccess) {
+                    this.requestLocationUpdates(
+                        LocationManager.GPS_PROVIDER,
+                        1000,
+                        0f,
+                        locationListener
+                    )
+                }
+                if (hasCoarseAccess) {
+                    this.requestLocationUpdates(
+                        LocationManager.NETWORK_PROVIDER,
+                        1000,
+                        0f,
+                        locationListener
+                    )
+                }
+            }?.onFailure {
+                Log.e("SearchableItemVM", "Failed to start location updates", it)
+            }
+    }
+
+    fun stopHeadingUpdates(context: Context) {
+        context.getSystemService<SensorManager>()?.unregisterListener(sensorListener)
+
+        // remove this if viewmodels are actually destroyed (I don't know)
+        trueNorthHeading.value = null
+    }
+
+    fun stopLocationUpdates(context: Context) {
+        context.getSystemService<LocationManager>()?.removeUpdates(locationListener)
+
+        // also applies to this
+        userLocation.value = null
+        geomagneticField = null
+    }
```

Comment:
This works, but the more elegant solution (imho) would be to only expose two methods `fun getUserLocation(context: Context): Flow<Location?>`/`getUserHeading(context: Context): Flow<Float?>` and then use a [`callbackFlow`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/callback-flow.html) to manage subscription to the location service. Then you could simply use [`Flow.collectAsStateWithLifecycle`](https://developer.android.com/reference/kotlin/androidx/lifecycle/compose/package-summary#(kotlinx.coroutines.flow.StateFlow).collectAsStateWithLifecycle(androidx.lifecycle.Lifecycle,androidx.lifecycle.Lifecycle.State,kotlin.coroutines.CoroutineContext)). This would also eliminate the need for a `repeatOnLifecycle` (see my comments in `LocationItem.kt`).

Reply:
In case of using a callback flow, do I still need to specify `lifecycleOwner` or does it default to the correct one?

Reply:
- You use a `callbackFlow` to make sure that the listener is correctly installed when the flow is started and removed when the flow is stopped
- You use `collectAsStateWithLifecycle`. This automatically starts the flow when the given lifecycle state (resumed by default) is reached and terminates it when the lifecycle falls below that state.
- When the flow is terminated, the `awaitClose` lambda is called. Here you unregister the listener
- Once the activity is resumed again, the flow is restarted and the code inside `callbackFlow` runs again

`collectAsStateWithLifecycle` uses the `LocalLifecycleOwner` by default which is the current activity, so no need to specify it further.

Use it like this in your composition:

```kt

val userLocation by remember(context) { viewModel.getUserLocation(context) }.collectAsStateWithLifecycle()

```

It's important to `remember` the result of `getUserLocation` because it is a function call if you didn't remember it, you would create a new flow object on every recomposition.

Reply:
Okay, because I just had `val userHeading by viewModel.getUserHeading(context).collectAsStateWithLifecycle(null)` and it wouldn't work anymore. I appreciate the guidance.

Reply:
Also, would you advise against using `trySendBlocking`? I always have the idea that blocking calls in UI-code causes issues, but nonblocking calls may need a buffer for unread values

Reply:
Whatever, the nonblocking variant seems to work fine without any buffers.